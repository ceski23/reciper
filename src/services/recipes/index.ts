export interface RecipeIngredient {
  text: string;
  group?: string;
}

export interface RecipeInstruction {
  text: string;
  group?: string;
}

export interface Recipe {
  id: string;
  name: string;
  image?: string;
  description?: string;
  ingredients: RecipeIngredient[];
  instructions: RecipeInstruction[];
  prepTime?: number;
  color?: string;
  url?: string;
  rating?: number;
  calories?: number;
  tags: string[];
  servings?: number;
}

// eslint-disable-next-line arrow-body-style
export const isValidRecipe = (recipe: Partial<Recipe>): recipe is Recipe => {
  // Has ID
  if (!recipe.id) return false;

  // Non-empty name
  if (!recipe.name) return false;

  // Has ingredients
  if (!recipe.ingredients || recipe.ingredients.length <= 0) return false;
  if (!recipe.ingredients[0].text) return false;

  // Has instructions
  if (!recipe.instructions || recipe.instructions.length <= 0) return false;
  if (!recipe.instructions[0].text) return false;

  // Has tags array
  if (!recipe.tags) return false;

  return true;
};
