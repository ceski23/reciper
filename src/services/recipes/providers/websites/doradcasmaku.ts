/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/doradca_smaku.png';

import { Recipe, RecipeIngredient } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import { removeEmpty } from 'utils/objects';

export const DoradcaSmakuProvider: Provider = (() => {
  const name = 'Doradca smaku';
  const url = 'https://www.doradcasmaku.pl/';

  const parseIngredients = (doc: Document) => {
    const ingredientsGroups = doc.querySelectorAll('.recipe_ingredients_inner > .ingredients_group');
    const ingredients: RecipeIngredient[] = [];

    for (const ingredientsGroup of ingredientsGroups) {
      const group = ingredientsGroup.querySelector('.ingredients_group_title')?.textContent?.trim();

      const groupIngredients = Array
        .from(ingredientsGroup.querySelectorAll('.ingredients_list > li'))
        .map((elem) => getTextFromNode(elem).trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => removeEmpty({ text, group }));

      ingredients.push(...groupIngredients);
    }

    return ingredients;
  };

  const parseImages = (doc: Document) => {
    const imageElements = doc.querySelectorAll('.recipe_item .recipe_gallery img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('data-src'))
      .filter(nonNullable);

    return images;
  };

  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    const recipeImageElem = doc.querySelector('.recipe_item .recipe_gallery .img_wrapper img');
    const image = recipeImageElem?.getAttribute('data-src') ?? undefined;

    const instructionsElements = doc.querySelectorAll('.recipe-description__content > *');

    const instructions = Array
      .from(instructionsElements)
      .map(getTextFromNode)
      .filter(nonNullable)
      .map((text) => ({ text }));

    const tags = Array
      .from(doc.querySelectorAll('.recipe_tags > a'))
      .map(getTextFromNode)
      .filter(nonNullable)
      .map((text) => text.toLocaleLowerCase());

    const uniqueTags = Array.from(new Set([...tags, ...(data.tags ?? [])]));

    const partial: Partial<Recipe> = {
      image,
      instructions,
      ingredients: parseIngredients(doc),
      tags: uniqueTags,
      gallery: parseImages(doc),
    };

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
