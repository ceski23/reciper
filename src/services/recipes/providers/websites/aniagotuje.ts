import icon from 'assets/providers/ania_gotuje.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { colorExtractor } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { isElementNode, nonNullable } from 'utils/guards';
import appLogger from 'utils/logger';
import { removeEmpty } from 'utils/objects';

const log = appLogger.extend('scrapper:aniagotuje');

export const AniaGotujeProvider: Provider = (() => {
  const name = 'AniaGotuje.pl';
  const url = 'https://aniagotuje.pl/';

  const parseIngredients = (doc: Document) => {
    log('parsing ingredients...');

    const root = doc.querySelector('[itemprop="recipeInstructions"] > .ads-slot-article + div');
    const ingredientsHeaders = root?.querySelectorAll<HTMLParagraphElement>('.ing-header');
    const ingredients: RecipeIngredient[] = [];

    // Single recipe
    if (ingredientsHeaders?.length === 0) {
      log('found single ingredients group');

      const groupItemsElement = root?.querySelectorAll('.recipe-ing-list [itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text }));

      ingredients.push(...groupItems);

      return ingredients;
    }

    log('found multiple ingredients groups');

    // Recipe with sub-recipes
    ingredientsHeaders?.forEach((ingHeader) => {
      const groupTitle = ingHeader.textContent?.trim();

      const groupItemsElement = ingHeader.nextElementSibling?.querySelectorAll('[itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text, group: groupTitle }));

      ingredients.push(...groupItems);
    });

    log('parsed ingredients:', ingredients);

    return ingredients;
  };

  const parseInstructions = (doc: Document) => {
    log('parsing instructions...');

    const instructions: RecipeInstruction[] = [];
    let currentElement: Node | Element | null = doc.querySelector('[itemprop="recipeInstructions"] .recipe-ing-list + h2');
    let header: string | undefined;

    while (currentElement !== null && currentElement !== undefined) {
      if (
        isElementNode(currentElement)
        && ['H1', 'H2', 'H3', 'H4'].includes(currentElement.tagName)
      ) {
        header = currentElement.textContent?.trim();
      } else {
        const text = getTextFromNode(currentElement);

        if (text !== null && text.length > 0) {
          instructions.push({ text, group: header });
        }
      }

      currentElement = currentElement.nextSibling;
    }

    log('parsed instructions:', instructions);

    return instructions;
  };

  const parseImages = (doc: Document) => {
    log('looking for images');

    const imageElements = doc.querySelectorAll('[itemprop="recipeInstructions"] img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('data-src'))
      .filter(nonNullable);

    log('found images:', images);

    return images;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    log('scrapping recipe...');

    const data = await microdataScrapper(doc);
    log('data from microdata:', data);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const recipeName = doc.querySelector('.article-content [itemprop="name"]')?.textContent?.trim();

    const partial: Partial<Recipe> = {
      color,
      name: recipeName,
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
      gallery: parseImages(doc),
    };

    log('additional data:', partial);

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
