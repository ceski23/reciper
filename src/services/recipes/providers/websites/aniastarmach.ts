import icon from 'assets/providers/ania_starmach.png';

import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import { colorExtractor } from 'services/recipes/providers/utils';

export const AniaStarmachProvider: Provider = (() => {
  const name = 'Ania Starmach';
  const url = 'https://aniastarmach.pl/';

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const tagsElements = doc.querySelectorAll('.tags-wrapper a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(Boolean) as string[];

    return {
      ...data,
      color,
      tags,
    };
  };

  return {
    name, url, scrapper, icon,
  };
})();
