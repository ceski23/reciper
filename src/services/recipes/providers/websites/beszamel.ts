import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import icon from 'assets/provider_icons/beszamel.svg';
import { colorExtractor } from 'services/recipes/providers/utils';

export const BeszamelProvider: Provider = (() => {
  const name = 'Beszamel';
  const url = 'https://beszamel.se.pl/';

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    return {
      ...data,
      color,
    };
  };

  return {
    name, url, scrapper, icon,
  };
})();
