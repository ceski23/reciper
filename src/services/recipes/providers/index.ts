/* eslint-disable import/no-cycle */
import fetch from 'cross-fetch';
import { nanoid } from 'nanoid';

import { isValidRecipe, Recipe } from 'services/recipes';
import { DefaultProvider } from 'services/recipes/providers/default';
import { AniaGotujeProvider } from 'services/recipes/providers/websites/aniagotuje';
import { AniaStarmachProvider } from 'services/recipes/providers/websites/aniastarmach';
import { BeszamelProvider } from 'services/recipes/providers/websites/beszamel';
import { DoradcaSmakuProvider } from 'services/recipes/providers/websites/doradcasmaku';
import { KwestiaSmakuProvider } from 'services/recipes/providers/websites/kwestiasmaku';

import appLogger from 'utils/logger';
import { baseUrl } from 'utils/url';

// ############################################################################
export const PROVIDERS = [
  KwestiaSmakuProvider,
  BeszamelProvider,
  AniaStarmachProvider,
  DoradcaSmakuProvider,
  AniaGotujeProvider,
];
// ############################################################################

export type RecipeScrapper = (doc: Document) => Promise<Partial<Recipe>>;

export interface Provider {
  name?: string;
  url?: string;
  icon?: string;
  scrapper: RecipeScrapper;
}

export const chooseProvider = (url: string): Provider => {
  const choosenProvider = PROVIDERS.find((provider) => provider.url === baseUrl(url));
  return choosenProvider ?? DefaultProvider;
};

export class InvalidRecipeError extends Error {}

const log = appLogger.extend('scrapper');

export const scrapeRecipe = async (url: string): Promise<Recipe> => {
  try {
    const recipeUrl = new URL(url);
    log('scrapping recipe at url: %s', recipeUrl);

    const targetUrl = (import.meta.env.VITE_CORS_PROXY as string ?? '') + encodeURIComponent(recipeUrl.toString());
    const data = await fetch(targetUrl).then((res) => res.text());

    const parser = new DOMParser();
    const doc = parser.parseFromString(data, 'text/html');

    const provider = chooseProvider(url);
    const partialRecipe = await provider.scrapper(doc);

    const recipe = {
      ...partialRecipe,
      id: nanoid(),
      url,
    };

    if (isValidRecipe(recipe)) {
      log('parsed recipe: %O', recipe);
      return recipe;
    }
    throw Error('Invalid recipe');
  } catch (error) {
    const anyError = error as Error;

    log(anyError);
    throw new InvalidRecipeError(anyError.message);
  }
};
