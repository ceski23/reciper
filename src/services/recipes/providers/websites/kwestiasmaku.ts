/* eslint-disable @typescript-eslint/no-loop-func */
/* eslint-disable @typescript-eslint/brace-style */
/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/kwestia_smaku.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { getColorFromImage } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import appLogger from 'utils/logger';
import { removeEmpty } from 'utils/objects';

const log = appLogger.extend('scrapper:kwestiasmaku');

export const KwestiaSmakuProvider: Provider = (() => {
  const name = 'Kwestia Smaku';
  const url = 'https://www.kwestiasmaku.com/';

  const parseIngredients = (doc: Document) => {
    const ingredientsRoot = doc.querySelector('.field-name-field-skladniki');
    if (!ingredientsRoot) return [];

    // Multiple ingredients groups
    if (ingredientsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const ingredients: RecipeIngredient[] = [];

      for (const element of ingredientsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeIngredients = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => elem.textContent?.trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          ingredients.push(...recipeIngredients);
        }
      }

      return ingredients;
    }

    // Signle ingredients group
    const ingredientsElements = ingredientsRoot.querySelectorAll('ul > li');
    const recipeIngredients = Array
      .from(ingredientsElements)
      .map((elem) => elem.textContent?.trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeIngredients;
  };

  const parseInstructions = (doc: Document) => {
    const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie');
    if (!instructionsRoot) return [];

    // Multiple instruction groups
    if (instructionsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const instructions: RecipeInstruction[] = [];

      for (const element of instructionsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeInstructions = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => getTextFromNode(elem).trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          instructions.push(...recipeInstructions);
        }
      }

      return instructions;
    }

    // Signle instructions group
    const instructionsElements = instructionsRoot.querySelectorAll('ul > li');
    const recipeInstructions = Array
      .from(instructionsElements)
      .map((elem) => getTextFromNode(elem).trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeInstructions;
  };

  const parseImages = (doc: Document) => {
    log('looking for images');

    const imageElements = doc.querySelectorAll('.view-zdjecia img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('src'))
      .filter(nonNullable);

    log('found images:', images);

    return images;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const microdata = await microdataScrapper(doc);
    const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim() || undefined;

    const tagsElements = doc.querySelectorAll('#node-przepis-full-group-kategorie a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(nonNullable);

    const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji');
    const servings = servingsElement?.textContent
      ? Number.parseInt(servingsElement.textContent, 10)
      : undefined;

    const color = microdata.image ? await getColorFromImage(microdata.image) : undefined;

    const partial: Partial<Recipe> = {
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
      color,
      description,
      tags,
      servings,
      gallery: parseImages(doc),
    };

    return Object.assign(microdata, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
