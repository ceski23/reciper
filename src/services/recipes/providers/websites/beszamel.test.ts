import { expect } from 'vitest';

import { Recipe } from 'services/recipes';
import { lemoniada } from 'services/recipes/providers/fixtures/beszamel';
import { BeszamelProvider } from 'services/recipes/providers/websites/beszamel';

const scrapeRecipe = async (data: string) => {
  const parser = new DOMParser();
  const doc = parser.parseFromString(data, 'text/html');
  const partialRecipe = await BeszamelProvider.scrapper(doc);
  partialRecipe.id = '1';

  return partialRecipe;
};

describe('scrapper', () => {
  it('should scrape valid recipe', async () => {
    const recipe = await scrapeRecipe(lemoniada);
    expect(recipe).toEqual<Recipe>({
      id: '1',
      ingredients: [
        { text: '2 pomarańcze' },
        { text: 'sok z 2 pomarańczy' },
        { text: '2 cytryny' },
        { text: 'sok z 1 cytryny' },
        { text: '2 limonki' },
        { text: 'sok z 2 limonek' },
        { text: 'syrop przygotowany z 2 litrów wody i 8 łyżek cukru' },
      ],
      instructions: [
        { text: 'Cytrusy umyć i pokroić w plasterki.' },
        { text: 'Wlać do dzbanka sok wyciśnięty z cytrusów i syrop cukrowy, a następnie dodać do niego pokrojone w plasterki owoce.' },
        { text: 'Całość wstawić na kilka godzin do lodówki.' },
        { text: 'Podawać z plasterkami owoców i lodem.' },
      ],
      name: 'Lemoniada cytrusowa - przepis na lemoniadę z cytryn, limonek i pomarańczy',
      tags: [
        'napoje',
        'napoje alkoholowe',
        'koktajle',
        'kompoty',
        'soki',
        'kawa i herbata',
      ],
      calories: undefined,
      color: '#7f7f7f',
      description: 'Lemoniada z limonek, cytryn i pomarańczy to bardzo prosty w przygotowaniu i niezwykle smaczny napój na upalne dni. Podajemy dobry przepis na lemoniadę cytrusową. Sprawdź, jak ją przygotować.',
      image: 'https://cdn.galleries.smcloud.net/t/photos/gf-EQGW-RBcL-arZc_lemoniada-cytrusowa-przepis-na-lemoniade-z-cytryn-limonek-i-pomaranczy.jpg',
      prepTime: 15,
      rating: undefined,
      servings: 4,
    });
  });
});
