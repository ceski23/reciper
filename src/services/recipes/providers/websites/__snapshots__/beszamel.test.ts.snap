// Vitest Snapshot v1

exports[`scrapper > should scrape valid recipe 1`] = `
{
  "color": "#7f7f7f",
  "description": "Lemoniada z limonek, cytryn i pomarańczy to bardzo prosty w przygotowaniu i niezwykle smaczny napój na upalne dni. Podajemy dobry przepis na lemoniadę cytrusową. Sprawdź, jak ją przygotować.",
  "gallery": [
    "https://cdn.galleries.smcloud.net/t/photos/gf-EQGW-RBcL-arZc_lemoniada-cytrusowa-przepis-na-lemoniade-z-cytryn-limonek-i-pomaranczy.jpg",
  ],
  "id": "1",
  "image": "https://cdn.galleries.smcloud.net/t/photos/gf-EQGW-RBcL-arZc_lemoniada-cytrusowa-przepis-na-lemoniade-z-cytryn-limonek-i-pomaranczy.jpg",
  "ingredients": [
    {
      "text": "2 pomarańcze",
    },
    {
      "text": "sok z 2 pomarańczy",
    },
    {
      "text": "2 cytryny",
    },
    {
      "text": "sok z 1 cytryny",
    },
    {
      "text": "2 limonki",
    },
    {
      "text": "sok z 2 limonek",
    },
    {
      "text": "syrop przygotowany z 2 litrów wody i 8 łyżek cukru",
    },
  ],
  "instructions": [
    {
      "text": "Cytrusy umyć i pokroić w plasterki.",
    },
    {
      "text": "Wlać do dzbanka sok wyciśnięty z cytrusów i syrop cukrowy, a następnie dodać do niego pokrojone w plasterki owoce.",
    },
    {
      "text": "Całość wstawić na kilka godzin do lodówki.",
    },
    {
      "text": "Podawać z plasterkami owoców i lodem.",
    },
  ],
  "name": "Lemoniada cytrusowa - przepis na lemoniadę z cytryn, limonek i pomarańczy",
  "prepTime": 15,
  "servings": 4,
  "tags": [
    "napoje",
    "napoje alkoholowe",
    "koktajle",
    "kompoty",
    "soki",
    "kawa i herbata",
  ],
}
`;

exports[`scrapper > should scrape valid recipe with ingredients groups 1`] = `
{
  "color": "#7f7f7f",
  "description": "Samodzielne przygotowanie tortu jest dość czasochłonne, ale wbrew pozorom nie takie trudne. Jeżeli po raz pierwszy zamierzasz zrobić tort - ten prosty i łatwy przepis jest właśnie dla ciebie. To nasz specjalny przepis na tort dla początkujących, który przy odrobinie wysiłku na pewno się uda.",
  "gallery": [
    "https://cdn.galleries.smcloud.net/t/photos/gf-eBME-h8Zw-WmRp_tort-dla-poczatkujacych.jpg",
  ],
  "id": "1",
  "image": "https://cdn.galleries.smcloud.net/t/photos/gf-eBME-h8Zw-WmRp_tort-dla-poczatkujacych.jpg",
  "ingredients": [
    {
      "group": "BISZKOPT:",
      "text": "12 łyżek mąki",
    },
    {
      "group": "BISZKOPT:",
      "text": "6 jajek",
    },
    {
      "group": "BISZKOPT:",
      "text": "1 szklanka cukru pudru",
    },
    {
      "group": "BISZKOPT:",
      "text": "1 łyżeczka masła i bułka tarta do formy",
    },
    {
      "group": "PONCZ:",
      "text": "1 łyżeczka czarnej herbaty",
    },
    {
      "group": "PONCZ:",
      "text": "4 łyżeczki cukru",
    },
    {
      "group": "PONCZ:",
      "text": "1 łyżka spirytusu",
    },
    {
      "group": "KREM:",
      "text": "1,5 kostki masła",
    },
    {
      "group": "KREM:",
      "text": "1 szklanka cukru pudru",
    },
    {
      "group": "KREM:",
      "text": "4 żółtka",
    },
    {
      "group": "KREM:",
      "text": "2 łyżki kakao",
    },
    {
      "group": "KREM:",
      "text": "1 łyżeczka mleka",
    },
    {
      "group": "DEKORACJA:",
      "text": "3-4 łyżki wiórków kokosowych",
    },
    {
      "group": "DEKORACJA:",
      "text": "12 obranych migdałów",
    },
  ],
  "instructions": [
    {
      "text": "Najpierw przygotuj ciasto - bazę tortu.",
    },
    {
      "text": "Oddziel żółtka od białek. Żółtka utrzyj z cukrem na jasną i puszystą masę. Białka ubij z odrobiną soli na sztywną pianę.",
    },
    {
      "text": "Delikatnie wymieszaj masę z żółtek z pianą białkową, a następnie dodawaj porcjami mąkę, mieszając szpatułką lub mikserem na najmniejszych obrotach.",
    },
    {
      "text": "Tak przygotowane ciasto przełóż do wysmarowanej masłem i posypanej bułką tartą tortownicy. Piecz około 40 minut, początkowo w 160 st. C, a po 15 minutach w 180 st. C.",
    },
    {
      "text": "Upieczone ciasto wyjmij z piekarnika i odstaw do ostygnięcia, wystudzone ciasto wyjmij z tortownicy i przekrój na 3 blaty.",
    },
    {
      "text": "Aby przygotować krem, miękkie masło utrzyj z cukrem pudrem na puszystą masę.",
    },
    {
      "text": "Dodaj kakao rozmieszane z mlekiem i - stale ucierając - po jednym żółtku. Dokładnie połącz składniki.",
    },
    {
      "text": "Herbatę zaparz w połowie szklanki wrzątku, posłódź i wymieszaj. Do wystudzonej dodaj spirytus. Nasącz blaty biszkoptowe.",
    },
    {
      "text": "Układaj blaty tortu jeden na drugim przekładając kremem. Kremem posmarować także boki i wierzch tortu.",
    },
    {
      "text": "Udekorować tartą czekoladą, sypkim kakao lub wiórkami kokosowymi i migdałami.",
    },
  ],
  "name": "Tort dla początkujących - łatwy przepis krok po kroku",
  "prepTime": 60,
  "servings": 4,
  "tags": [
    "ciasta okazjonalne",
    "ciasta regionalne",
    "sękacz",
    "pascha wielkanocna",
    "keks",
    "makowiec",
    "mazurki",
    "miodowniki",
    "pierniki",
    "torty",
    "babki",
  ],
}
`;
